std/math/powi.zig

```diff
@@ -0,0 +1,137 @@
+//  powi(x, 1)      = x for any x
+//  powi(nan, y)    = nan
+//  powi(x, nan)    = nan
+//  powi(+-0, y)    = +-inf for y an odd integer < 0
+//  powi(+-0, -inf) = +inf
+//  powi(+-0, +inf) = +0
+//  powi(+-0, y)    = +inf for finite y < 0 and not an odd integer
+//  powi(+-0, y)    = +-0 for y an odd integer > 0
+//  powi(+-0, y)    = +0 for finite y > 0 and not an odd integer
+//  powi(-1, +-inf) = 1
+//  powi(x, +inf)   = +inf for |x| > 1
+//  powi(x, -inf)   = +0 for |x| > 1
+//  powi(x, +inf)   = +0 for |x| < 1
+//  powi(x, -inf)   = +inf for |x| < 1
+//  powi(+inf, y)   = +inf for y > 0
+//  powi(+inf, y)   = +0 for y < 0
+//  powi(-inf, y)   = powi(-0, -y)
+//  powi(x, y)      = nan for finite x < 0 and finite non-integer y
+
+const builtin = @import("builtin");
+const std = @import("../index.zig");
+const math = std.math;
+const assert = std.debug.assert;
+const assertError = std.debug.assertError;
+
+inline fn checkFlow(comptime T: type, comptime info: builtin.TypeInfo, x: T, y: T) (error{Overflow, Underflow}!void) {
+    const bit_size = @sizeOf(T) * 8;
+
+    if (info.Int.is_signed) {
+        if (x != 1 and x != -1) {
+            if (x > 0 and y >= bit_size - 1) {
+                return error.Overflow;
+            } else if (x < 0 and y > bit_size - 1) {
+                return error.Underflow;
+            }
+        }
+    } else {
+        if (x != 1 and y >= bit_size) {
+            return error.Overflow;
+        }
+    }
+}
+
+// This implementation is based on that from the rust stlib
+pub fn safePowi(comptime T: type, x: T, y: T) !T {
+    const info = @typeInfo(T);
+
+    switch (info) {
+        builtin.TypeId.Int => {
+            try checkFlow(T, info, x, y);
+
+            var base = x;
+            var exp = y;
+            var acc: T = 1;
+
+            while (exp > 1) {
+                if (exp & 1 == 1) {
+                    if (@mulWithOverflow(T, acc, base, &acc)) {
+                        if (x > 0) {
+                            return error.Overflow;
+                        } else {
+                            return error.Underflow;
+                        }
+                    }
+                }
+
+                exp >>= 1;
+
+                if (@mulWithOverflow(T, base, base, &base)) {
+                    if (x > 0) {
+                        return error.Overflow;
+                    } else {
+                        return error.Underflow;
+                    }
+                }
+            }
+
+            if (exp == 1) {
+                if (@mulWithOverflow(T, acc, base, &acc)) {
+                    if (x > 0) {
+                        return error.Overflow;
+                    } else {
+                        return error.Underflow;
+                    }
+                }
+            }
+
+            return acc;
+        },
+        else => @compileError("expected integer type got " ++ @typeName(T)),
+    }
+}
+
+pub fn powi(comptime T: type, x: T, y: T) T {
```

Comment:
What do you think about having `std.math.pow` support floats and integers, and then `std.math.powi` is the safe one that has a potential error?

I'd be ok with either way. On one hand having `std.math.pow` support integers seems clean; on the other hand, it might be nice to have different functions to point out the fact that one of them is safe (floats) and one is potentially unsafe (ints). Your call.